river/tree/hoeffding_adaptive_tree_classifier.py

```diff
@@ -174,68 +177,50 @@ def learn_one(self, x, y, *, sample_weight=1.0):
     # Override HoeffdingTreeClassifier
     def predict_proba_one(self, x):
         proba = {c: 0.0 for c in self.classes}
-        if self._tree_root is not None:
-            found_nodes = self._filter_instance_to_leaves(x, None, -1)
-            for fn in found_nodes:
-                # parent_branch == -999 means that the node is the root of an alternate tree.
-                # In other words, the alternate tree is a single leaf. It is probably not accurate
-                # enough to be used to predict, so skip it
-                if fn.parent_branch != -999:
-                    leaf_node = fn.node
-                    if leaf_node is None:
-                        leaf_node = fn.parent
-                    dist = leaf_node.leaf_prediction(x, tree=self)
-                    # Option Tree prediction (of sorts): combine the response of all leaves reached
-                    # by the instance
-                    proba = add_dict_values(proba, dist, inplace=True)
+        if self._root is not None:
+            found_nodes = [self._root]
+            if isinstance(self._root, HTBranch):
+                found_nodes = self._root.traverse(x, until_leaf=True)
```

Comment:
How is the skipping of alternate tree roots enforced?

Reply:
You indeed have sharp eyes! Well, first some background:

The "option votes" are not part of the original HATC. In fact, a [recent paper](https://arxiv.org/abs/2010.08199) discusses that this and other changes were applied to the code in MOA, but they are not documented anywhere. In summary, "option votes" are not originally part of the theoretical HATC algorithm.

Some time ago, after trying to understand the HATC code for the zillionth time, I decided to add a comment concerning my understanding of the code. I did that because I kept forgetting what that and other portions did. So, _my understading_ is that alternate trees composed of a single leaf don't vote because they are "too fresh". But again, this is my interpretation of the code.

But well, if we are talking about unspecified behaviors, in my humble opinion, if concept drift was detected at a specific decision node, that poor new alternate tree might do a better job than the current subtree since it is being trained using only instances belonging to the new concept.

So, now we have a war of unspecified behaviors XD. In fact, I intend to add a parameter to switch on/off the "option votes" in the future. For simplicity, for now, I allowed every alternate tree to vote in this refactoring. Again, I mostly did that because of my personal bias regarding tree learning, but also because filtering alternate trees composed of a single leaf was a tricky endeavor.

What do you think? If needed, I can try to mimic the behavior in MOA. But if I'm not wrong, the mentioned paper discusses that sometimes the "option vote" is not beneficial. That case was also mentioned in #532.

In any case, my bad! I forgot to document that. It was lost among the other changes.

Reply:
This is indeed a tricky situation. I can ask the author of HATC ;-) for his opinion.

Reply:
That would be awesome!

Reply:
This falls outside of the scope for this PR so I am not going to block its merge based on it. If I get new information I will update the comment.